File: backend/services/user-service/src/main/java/com/training/platform/validation/PasswordStrengthValidator.java
package com.training.platform.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;

public class PasswordStrengthValidator implements ConstraintValidator<PasswordStrength, String> {

    @Override
    public void initialize(PasswordStrength constraintAnnotation) {
        // No initialization needed
    }

    @Override
    public boolean isValid(String password, ConstraintValidatorContext context) {
        if (password == null || password.isBlank()) {
            return true; // Let @NotBlank handle empty passwords
        }

        // Minimum length
        if (password.length() < 8) {
            return false;
        }

        // At least one uppercase letter
        boolean hasUppercase = password.chars().anyMatch(Character::isUpperCase);
        if (!hasUppercase) {
            return false;
        }

        // At least one lowercase letter
        boolean hasLowercase = password.chars().anyMatch(Character::isLowerCase);
        if (!hasLowercase) {
            return false;
        }

        // At least one digit
        boolean hasDigit = password.chars().anyMatch(Character::isDigit);
        if (!hasDigit) {
            return false;
        }

        // At least one special character
        boolean hasSpecialChar = password.chars()
                .anyMatch(ch -> !Character.isLetterOrDigit(ch));
        if (!hasSpecialChar) {
            return false;
        }

        return true;
    }
}

File: frontend/src/app/core/validators/password-strength.validator.ts
import { AbstractControl, ValidationErrors, ValidatorFn } from '@angular/forms';

/**
 * Validates password strength:
 * - At least 8 characters
 * - At least one uppercase letter
 * - At least one lowercase letter
 * - At least one digit
 * - At least one special character (!@#$%^&*()_+-=[]{}|;:,.<>?)
 */
export function passwordStrengthValidator(): ValidatorFn {
  return (control: AbstractControl): ValidationErrors | null => {
    if (!control.value) {
      return null; // Let required validator handle empty values
    }

    const value = control.value as string;
    const errors: ValidationErrors = {};

    // Minimum length
    if (value.length < 8) {
      errors['minLength'] = { requiredLength: 8, actualLength: value.length };
    }

    // At least one uppercase letter
    if (!/[A-Z]/.test(value)) {
      errors['hasUppercase'] = true;
    }

    // At least one lowercase letter
    if (!/[a-z]/.test(value)) {
      errors['hasLowercase'] = true;
    }

    // At least one digit
    if (!/[0-9]/.test(value)) {
      errors['hasDigit'] = true;
    }

    // At least one special character
    if (!/[!@#$%^&*()_+\-=\[\]{}|;:,.<>?]/.test(value)) {
      errors['hasSpecialChar'] = true;
    }

    return Object.keys(errors).length > 0 ? { passwordStrength: errors } : null;
  };
}
